src/sidebar/components/thread-list-omega.js

```diff
@@ -0,0 +1,246 @@
+
+    return () => {
+      scrollContainer.removeEventListener('scroll', updateScrollPosition);
+      window.removeEventListener('resize', updateScrollPosition);
+    };
+  }, []);
+
+  const topLevelThreads = thread.children;
+
+  // TODO - We could optimize here by memo-ing such that `topLevelThreadIds`
+  // only updates if the IDs did change.
+  const topLevelThreadIds = topLevelThreads.map(t => t.id);
+
+  // Effect to update measured thread heights when the set of threads
+  // changes.
+  useEffect(() => {
+    let newHeights = {};
+
+    for (let id of topLevelThreadIds) {
+      const threadElement = document.getElementById(id);
+      if (!threadElement) {
+        // Thread is currently off screen.
+        continue;
+      }
+      const height = getElementHeightWithMargins(threadElement);
+      if (height !== null) {
+        newHeights[id] = height;
+      }
+    }
+
+    setThreadHeights(heights => {
+      // Merge existing and new heights.
+      newHeights = Object.assign({}, heights, newHeights);
+
+      // Skip update if no heights actually changed.
+      if (shallowEqual(heights, newHeights)) {
+        return heights;
+      }
+      return newHeights;
+    });
+  }, [topLevelThreadIds]);
+
+  // TODO - We could optimize `calculateVisibleThreads` as follows:
+  //
+  // - Pass in thread IDs rather than threads (as result does not depend on
+  //   other properties of threads)
+  // - Memoize based on inputs (thread IDs, last-measured thread heights,
+  //   scroll position, window height)
+  const {
+    offscreenLowerHeight,
+    offscreenUpperHeight,
+    visibleThreads,
+  } = calculateVisibleThreads(
+    topLevelThreads,
+    threadHeights,
+    scrollPosition,
+    windowHeight
+  );
+
+  // Effect to scroll a particular thread into view. This is mainly used to
+  // scroll a newly created annotation into view.
+  useEffect(() => {
+    if (!scrollToId) {
+      return;
+    }
+
+    const threadIndex = topLevelThreads.findIndex(t => t.id === scrollToId);
+    if (threadIndex === -1) {
+      // Thread is not currently present.
+      //
+      // When `scrollToId` is set as a result of a `BEFORE_ANNOTATION_CREATED`
+      // event, the annotation is not always present in the _next_ render after
+      // that event is received, but in another render after that. Therefore
+      // we wait until the annotation appears before "consuming" the scroll-to-id.
+      return;
+    }
+
+    // Clear `scrollToId` so we don't scroll again after the next render.
+    setScrollToId(null);
+
+    const getThreadHeight = thread =>
+      threadHeights[thread.id] || THREAD_DEFAULT_HEIGHT;
+
+    const yOffset = topLevelThreads
+      .slice(0, threadIndex)
+      .reduce((total, thread) => total + getThreadHeight(thread), 0);
+
+    const scrollContainer = getScrollContainer();
+    scrollContainer.scrollTop = yOffset;
+
+  }, [scrollToId, topLevelThreads, threadHeights]);
```

Comment:
This effect depends on the current set of threads and the heights, but it bails out early if `scrollToId` isn't set, so most of the time this effect does nothing.